takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/Deploy.java

```diff
@@ -82,10 +93,15 @@ private void installProject(MavenProject project) throws MojoExecutionException
 
     deployRequest.setRepository(remoteRepository(project));
 
-    try {
-      repositorySystem.deploy(repositorySystemSession, deployRequest);
-    } catch (DeploymentException e) {
-      throw new MojoExecutionException(e.getMessage(), e);
+    if (!deployAtEnd) {
+      try {
+        repositorySystem.deploy(repositorySystemSession, deployRequest);
+      } catch (DeploymentException e) {
+        throw new MojoExecutionException(e.getMessage(), e);
+      }
+    } else {
+      DeployParticipant.deployAtEnd(mavenSession, deployRequest);
```

Comment:
That was the first thing I tried (participant was @Named @Singleton), but for some reason it was a singleton per-project, and a completely differrent instance at the point of lifecycleParticipant execution.
As far as I understood, it is due to projects and the overall build having separate classRealms and each had a different instance of that participant.